spark/src/main/scala/com/gu/tableversions/spark/SparkHiveMetastore.scala

```diff
@@ -0,0 +1,182 @@
+package com.gu.tableversions.spark
+
+import java.net.URI
+
+import cats.effect.Sync
+import cats.implicits._
+import com.gu.tableversions.core.Partition.{ColumnValue, PartitionColumn}
+import com.gu.tableversions.core._
+import com.gu.tableversions.metastore.Metastore.TableOperation
+import com.gu.tableversions.metastore.Metastore.TableOperation._
+import com.gu.tableversions.metastore.{Metastore, VersionPaths}
+import com.typesafe.scalalogging.LazyLogging
+import org.apache.spark.sql.SparkSession
+
+/**
+  * Concrete implementation of the Metastore API, using Spark and Hive APIs.
+  */
+class SparkHiveMetastore[F[_]](implicit spark: SparkSession, F: Sync[F]) extends Metastore[F] with LazyLogging {
+
+  import SparkHiveMetastore._
+  import spark.implicits._
+
+  override def currentVersion(table: TableName): F[TableVersion] = {
+
+    val partitionedTableVersion: F[TableVersion] = for {
+      partitions <- listPartitions(table)
+      partitionLocations <- partitions.map { partition =>
```

Comment:
A couple of opportunities for `traverse` here and below in `def update`:

```scala
.map(someApplicativeAction).sequence === .traverse(someApplicativeAction)
```

In the case of `def update`, `traverse_` (note the underscore) captures the `.void` bit too: "I don't care about the result, just sequence all of these actions for their effects"

I've also started to find that when I'm reifying control flow with `IO` and `traverse` and `flatMap`, it's a signal that pushes me towards fs2. We're perhaps not there yet, but maybe it's not too far off! 😁

Reply:
> A couple of opportunities for `traverse` here and below in `def update`
[...[
> In the case of `def update`, `traverse_` (note the underscore) captures the `.void` bit too: "I don't care about the result, just sequence all of these actions for their effects"

Ahhhh nice thanks! 🙌

> I've also started to find that when I'm reifying control flow with `IO` and `traverse` and `flatMap`, it's a signal that pushes me towards fs2. We're perhaps not there yet, but maybe it's not too far off! 😁

Yes, I did find myself missing fs2 while writing this.